pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,303 @@
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	osName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	teamID, err := strconv.Atoi(c.TeamID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid TeamID, details = %v", err),
+		}
+	}
+
+	var planID int = 0
+	plans, _, err := client.Plans.List(teamID)
+	for _, plan := range plans {
+		if plan.Name == c.Plan {
+			planID = plan.ID
+			break
+		}
+	}
+
+	serverCreateRequest := cherrygo.CreateServer{
+		ProjectID:   c.ProjectID,
+		Hostname:    machine.Name,
+		Image:       osName,
+		Region:      c.Location,
+		SSHKeys:     []string{},
+		IPAddresses: []string{},
+		PlanID:      strconv.Itoa(planID),
+		UserData:    base64.StdEncoding.EncodeToString([]byte(userdata)),
+	}
+
+	server, _, err := client.Server.Create(c.ProjectID, &serverCreateRequest)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to create server = %v", err),
+		}
+	}
+
+	return &cherryServer{server: server}, nil
+}
+
+type cherryServer struct {
+	server cherrygo.Server
+}
+
+func (s *cherryServer) Name() string {
+	return s.server.Hostname
+}
+
+func (s *cherryServer) ID() string {
+	return strconv.Itoa(s.server.ID)
+}
+
+func (s *cherryServer) Addresses() []string {
+	var addresses []string
+	for _, ip := range s.server.IPAddresses {
+		addresses = append(addresses, ip.Address)
+	}
+	return addresses
+}
+
+func (s *cherryServer) Status() instance.Status {
+	if s.server.State != "active" {
+		return instance.StatusCreating
+	}
+	return instance.StatusRunning
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	servers, _, err := client.Servers.List(c.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, server := range servers {
+		if server.Hostname == machine.Name {
```

Comment:
Is it safe to use the hostname as the mechanism for getting servers? Is it possible to happen that we have two servers with the same name? If yes, is this the case we need to be worried about?

Reply:
Actually you can have 2 cherry servers with the same hostname. The fact that they don't support labels/tags make it tricky to fetch servers. What would you recommend ?